Add reducers and actions

CarlosZ96 opened the pull request:
A state part for books with an array of books (initial state: empty array), reducer that adds a book, and reducer that removes an added book. Added a status slice for categories with an array of categories (initial state: empty array) and a reducer that checks for status and always returns "Under Construction" (initial state should check for that string).
The structure of the folders followed:
/src
--- |---/components
----|---/redux
 ------|--- /books
------------| booksSlice.js
-------|--- /categories
------------| categoriesSlice.js
 ------| store.js
---| index.js

CarlosZ96 committed:
Feature creating files and folders:

  Structure: |---/redux
        |--- /books
                | booksSlice.js
        |--- /categories
                | categoriesSlice.js
        | store.js

CarlosZ96 committed:
Feature Final:

A state part for books with an array of books (initial state: empty array), reducer that adds a book, and reducer that removes an added book. Added a status slice for categories with an array of categories (initial state: empty array) and a reducer that checks for status and always returns "Under Construction" (initial state should check for that string).
Structure of the folders followed:
/src
     |---/components
     |---/redux
         |--- /books
                 | booksSlice.js
         |--- /categories
                 | categoriesSlice.js
         | store.js
     | index.js

CarlosZ96 committed:
Fix Linters

Fix Linters

shahadat3669 commented on src/redux/store.js:
```diff
@@ -0,0 +1,12 @@
+import { configureStore } from '@reduxjs/toolkit';
```
- [ ] 👍 You're on the right track using `@reduxjs/toolkit` for your project! However, I've noticed that while you're using `@reduxjs/toolkit`, the necessary packages like `react-redux` and `@reduxjs/toolkit` aren't included in your `package.json` file. For a smooth setup, I kindly recommend configuring your project properly as outlined in the [documentation](https://github.com/microverseinc/curriculum-react-redux/blob/main/bookstore/add_reducers_an_actions_v2.md#:~:text=Add%20Redux%20Toolkit%20(npm%20install%20react%2Dredux%20%40reduxjs/toolkit).). This will involve adding the required packages to your `package.json` file.

By following the documentation closely and ensuring the correct package configuration, you'll create a solid foundation for your Redux implementation. Keep up the great work and thank you for your attention to detail! 🌟📦

shahadat3669 commented on src/redux/categories/categoriesSlice.js:
```diff
@@ -0,0 +1,19 @@
+    categories: [],
+  },
+  reducers: {
+    checkStatus: (state) => {
+      state.categories = 'Under construction';
+    },
```
👍 You've been consistently returning 'Under construction' from your `checkStatus` reducer, which demonstrates your dedication to maintaining this behavior. Your commitment is truly evident. However, I'd like to highlight a specific requirement stated in the [project documentation](https://github.com/microverseinc/curriculum-react-redux/blob/main/bookstore/add_reducers_an_actions_v2.md#:~:text=Includes%20a%20reducer%20that%20checks%20the%20status%20and%20always%20returns%20%22Under%20construction%22%20(the%20initial%20state%20should%20check%20to%20that%20string)). As per the project requirement, it's essential that the reducer not only returns "Under construction" consistently but also **checks the status** to ensure this behavior. Presently, your implementation directly updates the status, rather than checking it. To align with the requirement, I recommend updating your code to include a condition that checks the status and then updates the `state.categories` to 'Under construction'. Here's one possible approach:

```suggestion
    checkStatus: (state, action) => {
      if (action.payload === "Under construction") {
        state.categories = "Under construction";
      }
    },
```

This way, you'll fulfill the requirement of checking the status before making the update. Your dedication to these specifications will undoubtedly enhance the quality of your project. Keep up the excellent work! 🌟🚧

shahadat3669 commented on src/redux/books/booksSlice.js:
```diff
@@ -0,0 +1,18 @@
+    addBook: (state, action) => {
+      state.books.push(action.payload);
+    },
+    removeBook: (state, action) => {
+      state.books.pop(action.payload);
+    },
```
- [ ] Well! 👍 I see that you're utilizing the `array.pop()` method to remove a specific book. While you're on the right track, there's a slight misunderstanding regarding the usage of the `pop()` method in JavaScript. To clarify, the `pop()` method is intended to remove the **last** element from an array and return it. However, it appears you're attempting to use `pop(action.payload)` where `action.payload` likely represents a value or variable you intend to remove from the array. In this context, the `pop()` method isn't suitable.

- [ ] If you're looking to remove an element from an array based on its value, I recommend using the `array.filter()` method. Here's how you can achieve that:

```javascript
// Example code to remove an element by its value using array.filter()
state.books = state.books.filter(book => book.id !== action.id);
```

By implementing the `array.filter()` method with this structure, you'll be able to remove the specific book from the array as intended. Your understanding of array methods is growing, and these refinements will contribute to the accuracy of your project. Keep up the excellent work! 🌟📚

shahadat3669 commented on src/redux/books/booksSlice.js:
```diff
@@ -0,0 +1,18 @@
+    },
+  },
+});
+
```
- [ ] You're doing a commendable job by exporting reducers within the `categoriesSlice`! This approach keeps your code well-organized and maintainable. To maintain consistency and prepare for future milestones, I'd recommend adopting a similar approach for exporting `addBook` and `removeBook` as actions within their respective slices. By exporting these actions, you're setting a solid foundation for your project's scalability. As your project evolves, having these actions available in different components will facilitate smoother development and enhance code reusability.

Your foresight in these matters will certainly pay off as your project grows. Keep up the fantastic work, and thank you for your commitment to a structured and efficient codebase! 🌟📦

shahadat3669 reviewed (changes requested):
Hi @CarlosZ96, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.

<img  src="https://media2.giphy.com/media/m3864rBwwBTKMAbpn3/giphy.gif" />


## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Has a slice of state for `books`
✅ Has a slice of state for `categories`
✅ Structured files using the [ducks pattern](https://github.com/erikras/ducks-modular-redux)


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Your openness to these suggestions demonstrates your dedication to adopting best practices in version control and collaboration. Keep up the great work, and thank you for your attention to these important details! 🌟📝

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        <img src="https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white" alt="GitHub">
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        <img src="https://img.shields.io/badge/linkedin-%230077B5.svg?style=for-the-badge&logo=linkedin&logoColor=white" alt="LinkedIn">
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        <img src="https://img.shields.io/badge/Slack-4A154B?style=for-the-badge&logo=slack&logoColor=white" alt="Slack">
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        <img src="https://img.shields.io/badge/Gmail-D14836?style=for-the-badge&logo=gmail&logoColor=white" alt="Email">
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

CarlosZ96 committed:
Feature:

Editing @reduxjs/toolkit and adding configurations for the package Json, edited checkStatus and removeBook.

DuaneDave reviewed (approved):
Hi @CarlosZ96 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✔️ Clean and clear implementation of reducer and actions

## Optional suggestions
N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
